### PR TITLE
[ZeRO-1] add args for save master weights and fp32 grad acc

### PR DESCRIPTION
**Pull request by hgt312**
- add args for save master weights/fp32 grad acc
- also doc strings
- change saved master weights format from list to dict (it's for DCP to transform the dict to a useable format)

### CODE REVIEW

File: torch_xla/distributed/zero_redundancy_optimizer.py [hunk 1]
```diff
@@ -57,11 +69,13 @@ def __init__(
...
       max_norm: Optional[float] = None,
       pin_layout: bool = True,
-      use_grad_acc_hook: bool = False,
       sharding_groups: Optional[Any] = None,
       grad_norm_groups: Optional[Any] = None,
       lazy_init: bool = False,
       coalesce_cc: bool = False,
+      use_grad_acc_hook: bool = False,
+      save_master_weights: bool = False,
+      higher_cc_precision: bool = False,
```
Comment (aws-rhsoln) [inline torch_xla/distributed/zero_redundancy_optimizer.py:78]: why do we need this? If once the optimizer dtype is set, shouldn't all ops in optimizer use the same dtype?

File: torch_xla/distributed/zero_redundancy_optimizer.py [hunk 2]
```diff
@@ -411,7 +435,7 @@ def allgather_weights_and_update_full_parameter(self):
           if param.grad is not None or (self.use_grad_acc_hook and
                                         hasattr(shard, 'main_grad')):
             padded_param = padded_params[index]
-            if param.dtype != self.optimizer_dtype:
+            if padded_param != param.dtype:
```
Comment (aws-rhsoln) [inline torch_xla/distributed/zero_redundancy_optimizer.py:438]: bug? We are comparing param to dtype

File: torch_xla/distributed/zero_redundancy_optimizer.py (hunk 1 above)
Comment (hgt312) [reply torch_xla/distributed/zero_redundancy_optimizer.py:78]: it controls the cc ops dtype, 2 reason to have this:
- without this, we see performance drop from tutorial, because it's from bf16/bf16 to fp32/fp32
- Haozheng's gpu compatible branch need this, we make it a config in nxd

[reduce-scatter dtype]/[all-gather dtype]
| use_grad_acc_hook/higher_cc_precision | True      | False     |
| ------------------------------------- | --------- | --------- |
| True                                  | fp32/fp32 | fp32/bf16 |
| False                                 | fp32/fp32 | bf16/bf16 |

Comment (HahTK) [reply torch_xla/distributed/zero_redundancy_optimizer.py:78]: For production version of this we need to ensure the feature matches the expected use cases. I think there are some gaps. Please refer to use case below. 

Also let me know if I missed anything

Checkpoint save : 
+++++++++++
1. Default use case : Save Master Weights and Optimizer State. Do not save : Forward Weights. 
2. Backward compatibility : Save Forward Weights and Optimizer State. Do not save : Master Weights. 
3. There is no use case for saving all 3 master + forward + optimizer but also no reason to prevent it ? 

DEFAULT BEHAVIOR should be (1). 
WARNINGS : We should print warnings for use case (2) and (3).
  * (2) - warn user that this has accuracy issues if they use it to resume training.
  * (3) - warn user that they are wasting disk space and forward weights will be ignored on load.
NOTE : There may exist use cases where customer want (2) because they want to publish lower precision checkpoint for inference or fine-tuning etc. We recommend that be done using an offline conversion script. A convenience feature maybe to automatically dump an *extra* lower precision checkpoint only on the final checkpoint. Some users continue training even from "completed" training runs with new data so the final checkpoint still needs to have the full precision.

Checkpoint load : 
+++++++++++
1. Default use case : Load master weights and Optimizer. Broadcast that to forward weights and ignore forward weights if they were saved.
2. Backward Compatibility : If master weights does not exist, the run should *FAIL* by default. Then expose an option to force a training to resume from lower precision weights

DEFAULT BEHAVIOR should be (1) and FAIL on (2).
WARNINGS : We should inform the user that continuing training from low precision weights will affect accuracy, especially if there were multiple crashes close in time because the errors introduced by each save/restore will compound itself.  We should print the option needed to override the FAIL in the warning message.

File: torch_xla/distributed/zero_redundancy_optimizer.py [hunk 3]
```diff
@@ -394,6 +416,8 @@ def allgather_weights_and_update_full_parameter(self):
                 pin_layout=self.pin_layout,
                 groups=self.sharding_groups,
             )
+            if padded_param != param.dtype:
```
Comment (aws-rhsoln) [inline torch_xla/distributed/zero_redundancy_optimizer.py:419]: same bug?

Comment (hgt312) [reply torch_xla/distributed/zero_redundancy_optimizer.py]: nice catch..

File: torch_xla/distributed/zero_redundancy_optimizer.py (hunk 1 above)
Comment (hgt312) [reply torch_xla/distributed/zero_redundancy_optimizer.py:78]: agree with these. while these should be done in nxd/nnm level.

for this PR, do we want to use `True` as default for `save_master_weights` and `higher_cc_precision`?

Comment (HahTK) [reply torch_xla/distributed/zero_redundancy_optimizer.py:78]: Instead of duping code in NNM and NXD, I think the warnings should be in ZERO1. 
IMHO, the default should be to use save_master_weights = True but this means it is a change that must be documented in release notes. 

for `higher_cc_precision` this is a product decision.

**Review by aws-rhsoln — APPROVED** [review]

**Comment by jeffhataws** [comment]
@JackCaoG  @alanwaketan  this is followup PR to https://github.com/pytorch/xla/pull/6863 and https://github.com/pytorch/xla/pull/6862 . Will cherry-pick all of these for master soon.

**Comment by JackCaoG** [comment]
Ah, it is to your 2.1 branch. SG, I will just merge it then.
